imsim/diffraction.py

```diff
@@ -0,0 +1,166 @@
+"""Routines to implement diffraction due to the LSST spider.
+
+The core idea is taken from https://ntrs.nasa.gov/citations/19990094899
+For more details, see doc/diffraction.md (Statistical Aproach)
+"""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from collections.abc import Callable
+
+import numpy as np
+
+
+@dataclass
+class Geometry:
+    """Defining a 2D geometry consisting of circles and lines with a thickness.
+
+    Parameters
+    ----------
+    thick_lines : numpy array of shape (n,4) where each row has the form
+        [nx, ny, d, thickness]. (nx, ny) is the normal to the line and d is
+        such that d*(nx, ny) lies on the line.
+    circles : numpy array of shape (n,3) where each row has the form
+        [x, y, r]. (x,y) is the center of the circle, and r the radius.
+    """
+
+    thick_lines: np.ndarray
+    circles: np.ndarray
+
+
+LSST_SPIDER_GEOMETRY = Geometry(
+    thick_lines=np.array(
+        [
+            [1 / np.sqrt(2.0), 1 / np.sqrt(2.0), -0.4, 0.025],
+            [-1 / np.sqrt(2.0), 1 / np.sqrt(2.0), -0.4, 0.025],
+            [1 / np.sqrt(2.0), 1 / np.sqrt(2.0), 0.4, 0.025],
+            [-1 / np.sqrt(2.0), 1 / np.sqrt(2.0), 0.4, 0.025],
+        ]
+    ),
+    circles=np.array([[0.0, 0.0, 2.3905], [0.0, 0.0, 4.35], [0.0, 0.0, 2.4005]]),
+)
+
+
+def diffraction_kick(
+    pos: np.ndarray,
+    v_z: np.ndarray,
+    wavelength: np.ndarray,
+    geometry: Geometry,
+    distribution: Callable[[np.ndarray], np.ndarray],
+) -> np.ndarray:
+    """For a ray of photons with positions pos, wavelengths wavelength,
+    randomly generate diffraction angles and compute the resulting change of the
+    directions of the photons within the pupil plane.
+
+    Parameters
+    ----------
+    pos : 2d positions of the intersections of the rays with the pupil plane.
+    v_z : Component of the direction of the photons perpendicular to the pupil plane.
+    wavelength : Wavelength of the photons.
+    geometry : Geometry representing the 2d projection of the spider of a telescope
+        into the pupil plane.
+    distribution : Random number generator representing the distribution of the diffraction
+        angles (depending on the wavelength and distance to the geometry).
+        It should be a callable accepting an array for phi_star and returning the
+        random values for phi
+    """
+
+    d, n = directed_dist(geometry, pos)
+    d_tan_phi = distribution(phi_star(d, wavelength))
+    return d_tan_phi[:, None] * v_z[:, None] * n
+
+
+def phi_star(delta: np.ndarray, wavelength: np.ndarray) -> np.ndarray:
+    """According to https://ntrs.nasa.gov/citations/19990094899
```

Comment:
Add something to the doc string stating this is the standard deviation in radians for a  deflection.